Flaky test TestSingleBinaryWithMemberlistScaling

Error cleared on re-run.

```
=== RUN   TestSingleBinaryWithMemberlistScaling
10:10:51 Starting dynamodb
10:10:52 dynamodb: Initializing DynamoDB Local with the following configuration:
10:10:52 dynamodb: Port:	8000
10:10:52 dynamodb: InMemory:	true
10:10:52 dynamodb: DbPath:	null
10:10:52 dynamodb: SharedDb:	true
10:10:52 dynamodb: shouldDelayTransientStatuses:	false
10:10:52 dynamodb: CorsParams:	*
10:10:52 Ports for container: e2e-cortex-test-dynamodb Mapping: map[8000:50262]
10:10:52 Starting cortex-1
10:10:53 cortex-1: level=warn ts=2021-07-09T10:10:53.426880616Z caller=modules.go:325 msg="Worker address is empty in single binary mode.  Attempting automatic worker configuration.  If queries are unresponsive consider configuring the worker explicitly." address=127.0.0.1:9095
10:10:53 Ports for container: e2e-cortex-test-cortex-1 Mapping: map[80:50268 8000:50266 9095:50264]
10:10:53 cortex-1: level=warn ts=2021-07-09T10:10:53.612131437Z caller=lifecycler.go:237 msg="found an existing instance(s) with a problem in the ring, this instance cannot become ready until this problem is resolved. The /ring http endpoint on the distributor (or single binary) provides visibility into the ring." ring=ingester err="instance cortex-1 in state PENDING"
10:10:53 cortex-1: level=warn ts=2021-07-09T10:10:53.835476236Z caller=lifecycler.go:237 msg="found an existing instance(s) with a problem in the ring, this instance cannot become ready until this problem is resolved. The /ring http endpoint on the distributor (or single binary) provides visibility into the ring." ring=ingester err="instance cortex-1 in state PENDING"
10:10:54 cortex-1: level=warn ts=2021-07-09T10:10:54.264851716Z caller=lifecycler.go:237 msg="found an existing instance(s) with a problem in the ring, this instance cannot become ready until this problem is resolved. The /ring http endpoint on the distributor (or single binary) provides visibility into the ring." ring=ingester err="instance cortex-1 in state PENDING"
10:10:54 cortex-1: level=warn ts=2021-07-09T10:10:54.71687778Z caller=lifecycler.go:237 msg="found an existing instance(s) with a problem in the ring, this instance cannot become ready until this problem is resolved. The /ring http endpoint on the distributor (or single binary) provides visibility into the ring." ring=ingester err="instance cortex-1 in state JOINING"
10:10:55 cortex-1: level=warn ts=2021-07-09T10:10:55.158490013Z caller=lifecycler.go:237 msg="found an existing instance(s) with a problem in the ring, this instance cannot become ready until this problem is resolved. The /ring http endpoint on the distributor (or single binary) provides visibility into the ring." ring=ingester err="instance cortex-1 in state JOINING"
10:10:55 cortex-1: level=warn ts=2021-07-09T10:10:55.569024857Z caller=lifecycler.go:237 msg="found an existing instance(s) with a problem in the ring, this instance cannot become ready until this problem is resolved. The /ring http endpoint on the distributor (or single binary) provides visibility into the ring." ring=ingester err="instance cortex-1 in state JOINING"
10:10:56 cortex-1: level=warn ts=2021-07-09T10:10:56.058551291Z caller=lifecycler.go:237 msg="found an existing instance(s) with a problem in the ring, this instance cannot become ready until this problem is resolved. The /ring http endpoint on the distributor (or single binary) provides visibility into the ring." ring=ingester err="instance cortex-1 in state JOINING"
10:10:56 cortex-1: level=warn ts=2021-07-09T10:10:56.476213126Z caller=lifecycler.go:237 msg="found an existing instance(s) with a problem in the ring, this instance cannot become ready until this problem is resolved. The /ring http endpoint on the distributor (or single binary) provides visibility into the ring." ring=ingester err="instance cortex-1 in state JOINING"
10:10:56 cortex-1: level=warn ts=2021-07-09T10:10:56.888474892Z caller=lifecycler.go:237 msg="found an existing instance(s) with a problem in the ring, this instance cannot become ready until this problem is resolved. The /ring http endpoint on the distributor (or single binary) provides visibility into the ring." ring=ingester err="instance cortex-1 in state JOINING"
10:10:57 cortex-1: level=warn ts=2021-07-09T10:10:57.333036864Z caller=lifecycler.go:237 msg="found an existing instance(s) with a problem in the ring, this instance cannot become ready until this problem is resolved. The /ring http endpoint on the distributor (or single binary) provides visibility into the ring." ring=ingester err="instance cortex-1 in state JOINING"
10:10:57 cortex-1: level=warn ts=2021-07-09T10:10:57.743535009Z caller=lifecycler.go:237 msg="found an existing instance(s) with a problem in the ring, this instance cannot become ready until this problem is resolved. The /ring http endpoint on the distributor (or single binary) provides visibility into the ring." ring=ingester err="instance cortex-1 in state JOINING"
10:10:58 cortex-1: level=warn ts=2021-07-09T10:10:58.183343421Z caller=lifecycler.go:237 msg="found an existing instance(s) with a problem in the ring, this instance cannot become ready until this problem is resolved. The /ring http endpoint on the distributor (or single binary) provides visibility into the ring." ring=ingester err="instance cortex-1 in state JOINING"
    integration_memberlist_single_binary_test.go:210: cortex-3: cortex_ring_members=4.000000 memberlist_client_kv_store_value_tombstones=16.000000
    integration_memberlist_single_binary_test.go:210: cortex-1: cortex_ring_members=4.000000 memberlist_client_kv_store_value_tombstones=16.000000
    integration_memberlist_single_binary_test.go:210: cortex-2: cortex_ring_members=4.000000 memberlist_client_kv_store_value_tombstones=16.000000
    integration_memberlist_single_binary_test.go:210: cortex-3: cortex_ring_members=4.000000 memberlist_client_kv_store_value_tombstones=16.000000
    integration_memberlist_single_binary_test.go:210: cortex-1: cortex_ring_members=4.000000 memberlist_client_kv_store_value_tombstones=16.000000
    integration_memberlist_single_binary_test.go:210: cortex-2: cortex_ring_members=4.000000 memberlist_client_kv_store_value_tombstones=16.000000
    integration_memberlist_single_binary_test.go:210: cortex-3: cortex_ring_members=4.000000 memberlist_client_kv_store_value_tombstones=16.000000
    integration_memberlist_single_binary_test.go:210: cortex-1: cortex_ring_members=4.000000 memberlist_client_kv_store_value_tombstones=16.000000
    integration_memberlist_single_binary_test.go:210: cortex-2: cortex_ring_members=4.000000 memberlist_client_kv_store_value_tombstones=16.000000
    integration_memberlist_single_binary_test.go:210: cortex-3: cortex_ring_members=4.000000 memberlist_client_kv_store_value_tombstones=16.000000
    integration_memberlist_single_binary_test.go:210: cortex-1: cortex_ring_members=4.000000 memberlist_client_kv_store_value_tombstones=16.000000
    integration_memberlist_single_binary_test.go:210: cortex-2: cortex_ring_members=4.000000 memberlist_client_kv_store_value_tombstones=16.000000
    integration_memberlist_single_binary_test.go:210: cortex-3: cortex_ring_members=4.000000 memberlist_client_kv_store_value_tombstones=16.000000
    integration_memberlist_single_binary_test.go:210: cortex-1: cortex_ring_members=4.000000 memberlist_client_kv_store_value_tombstones=16.000000
    integration_memberlist_single_binary_test.go:210: cortex-2: cortex_ring_members=4.000000 memberlist_client_kv_store_value_tombstones=16.000000
    integration_memberlist_single_binary_test.go:210: cortex-3: cortex_ring_members=4.000000 memberlist_client_kv_store_value_tombstones=16.000000
    integration_memberlist_single_binary_test.go:210: cortex-1: cortex_ring_members=4.000000 memberlist_client_kv_store_value_tombstones=16.000000
    integration_memberlist_single_binary_test.go:210: cortex-2: cortex_ring_members=4.000000 memberlist_client_kv_store_value_tombstones=16.000000
    integration_memberlist_single_binary_test.go:210: cortex-3: cortex_ring_members=4.000000 memberlist_client_kv_store_value_tombstones=16.000000
    integration_memberlist_single_binary_test.go:210: cortex-1: cortex_ring_members=4.000000 memberlist_client_kv_store_value_tombstones=16.000000
    integration_memberlist_single_binary_test.go:210: cortex-2: cortex_ring_members=4.000000 memberlist_client_kv_store_value_tombstones=16.000000
    integration_memberlist_single_binary_test.go:210: cortex-3: cortex_ring_members=4.000000 memberlist_client_kv_store_value_tombstones=16.000000
    integration_memberlist_single_binary_test.go:210: cortex-1: cortex_ring_members=4.000000 memberlist_client_kv_store_value_tombstones=16.000000
    integration_memberlist_single_binary_test.go:210: cortex-2: cortex_ring_members=4.000000 memberlist_client_kv_store_value_tombstones=16.000000
    integration_memberlist_single_binary_test.go:210: cortex-3: cortex_ring_members=4.000000 memberlist_client_kv_store_value_tombstones=16.000000
    integration_memberlist_single_binary_test.go:210: cortex-1: cortex_ring_members=4.000000 memberlist_client_kv_store_value_tombstones=16.000000
    integration_memberlist_single_binary_test.go:210: cortex-2: cortex_ring_members=4.000000 memberlist_client_kv_store_value_tombstones=16.000000
    integration_memberlist_single_binary_test.go:210: cortex-3: cortex_ring_members=4.000000 memberlist_client_kv_store_value_tombstones=16.000000
    integration_memberlist_single_binary_test.go:210: cortex-1: cortex_ring_members=4.000000 memberlist_client_kv_store_value_tombstones=16.000000
    integration_memberlist_single_binary_test.go:210: cortex-2: cortex_ring_members=4.000000 memberlist_client_kv_store_value_tombstones=16.000000
    integration_memberlist_single_binary_test.go:210: cortex-3: cortex_ring_members=4.000000 memberlist_client_kv_store_value_tombstones=16.000000
    integration_memberlist_single_binary_test.go:203: 
        	Error Trace:	integration_memberlist_single_binary_test.go:203
        	Error:      	Condition never satisfied
        	Test:       	TestSingleBinaryWithMemberlistScaling
        	Messages:   	expected all instances to have 3.000000 ring members and 17.000000 tombstones
10:13:48 Killing cortex-3
10:13:48 cortex-2: level=warn ts=2021-07-09T10:13:48.990739601Z caller=tcp_transport.go:417 msg="TCPTransport: WriteTo failed" addr=192.168.0.5:8000 err="dial tcp 192.168.0.5:8000: connect: connection refused"
10:13:49 Killing cortex-2
10:13:49 Killing cortex-1
10:13:49 Killing dynamodb
--- FAIL: TestSingleBinaryWithMemberlistScaling (178.83s)
```

--- a/integration/integration_memberlist_single_binary_test.go
+++ b/integration/integration_memberlist_single_binary_test.go
@@ -197,6 +197,13 @@ func TestSingleBinaryWithMemberlistScaling(t *testing.T) {
 		c := instances[i]
 		instances = instances[:i]
 		stop.Go(func() error { return s.Stop(c) })
+
+		// TODO(#4360): Remove this when issue is resolved.
+		//   Wait until memberlist for all nodes has recognised the instance left.
+		//   This means that we will not gossip tombstones to leaving nodes.
+		for _, c := range instances {
+			require.NoError(t, c.WaitSumMetrics(e2e.Equals(float64(len(instances))), "memberlist_client_cluster_members_count"))
+		}
 	}
 	require.NoError(t, stop.Wait())
 

